### PR TITLE
Account for PayPal in claiming workflow

### PR DESCRIPTION
**Pull request by whit537**
Part of #4305, follows on #4424.

Brainstorm: let's account for this by automatically using their newly verified email as their PayPal, and telling them that if they want to change it they can.

- [x] ~~look into possible bug in `revive` where one participants gets the other's paypal linked~~ &hellip; bumped to #4437

### CODE REVIEW

**Comment by whit537** [comment]
I think we can do so much better than this page. I want it to feel like a consistent experience to onboard using this method.

screen shot 2017-04-27 at 2 53 32 pm

**Comment by whit537** [comment]
Another thing is that we can get to this page via an email that is already linked, in which case "is now connected" is the wrong message.

**Comment by whit537** [comment]
I can't worry about styling this page right now. I just need to set PayPal if needed and say as much in the UI. Styling will have to happen post-#4305.

**Comment by whit537** [comment]
Alright, nice and ugly.

screen shot 2017-04-27 at 7 05 16 pm

**Comment by whit537** [comment]
Rebased to pick up 47ad781406bbc5d1bec5e12f5fc8bcb08a950177 up in #4410, was 698640ae85c6900f2ee896ac3a28bf726c6b8ec4.

**Comment by whit537** [comment]
Rebased again to pick up Checkpoint 2 consolidation on #4305.

File: gratipay/models/participant/email.py [hunk 1]
```diff
@@ -233,31 +233,36 @@ def _set_primary_email(self, email, cursor):
...
+                return VERIFICATION_FAILED, None, None
             if (utcnow() - record.verification_start) > EMAIL_HASH_TIMEOUT:
-                return VERIFICATION_FAILED, None
+                return VERIFICATION_FAILED, None, None
             try:
+                paypal_updated = False
                 if packages:
                     self.finish_package_claims(cursor, nonce, *packages)
                 self.save_email_address(cursor, email)
+                if packages and not self.get_payout_routes(good_only=True):
```
Comment (dowski) [inline gratipay/models/participant/email.py]: What does this signify?

The negative boolean plus the `True` flag makes it hard to parse. Can you convert it to a positive condition?

File: gratipay/models/participant/email.py [hunk 2]
```diff
@@ -233,31 +233,36 @@ def _set_primary_email(self, email, cursor):
 
 
     def finish_email_verification(self, email, nonce):
-        """Given an email address and a nonce as strings, return a tuple:
-        (a ``VERIFICATION_*`` constant, and a list of packages for
-        ``VERIFICATION_SUCCEEDED`` or ``None`` otherwise)
+        """Given an email address and a nonce as strings, return a tuple: (a
+        ``VERIFICATION_*`` constant, a list of packages for
+        ``VERIFICATION_SUCCEEDED`` or ``None`` otherwise, and a boolean
+        indicating whether the participant's PayPal address was updated)
```
Comment (dowski) [inline gratipay/models/participant/email.py]: "or `None` indicating PayPal is not applicable." ... ? Something like that.

File: gratipay/models/participant/routes.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import braintree
+from psycopg2 import IntegrityError, errorcodes
+
+from ..exchange_route import ExchangeRoute
+from ...billing.instruments import CreditCard
+from ...utils import is_card_expiring
+
+
+class Routes(object):
+    """Participants have payment routes to get money into and out of Gratipay.
```
Comment (dowski) [inline gratipay/models/participant/routes.py]: Maybe add a note that this is a `Participant` mixin?

File: gratipay/models/participant/routes.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    @property
+    def has_payout_route(self):
+        """A boolean computed property, whether the participant has a known-working payout route.
+        """
+        return bool(self.get_payout_routes(good_only=True))
+
+
+    def get_payout_routes(self, good_only=False, cursor=None):
+        """Return a list of payout routes. If ``good_only`` evaluates to rue then only
```
Comment (dowski) [inline gratipay/models/participant/routes.py]: Typo: s/rue/True

File: gratipay/models/participant/routes.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        known-working payout routes are included.
+        """
+        out = []
+        for network in ('paypal',):
```
Comment (dowski) [inline gratipay/models/participant/routes.py]: List not tuple.

File: gratipay/models/participant/routes.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+import braintree
+from psycopg2 import IntegrityError, errorcodes
+
+from ..exchange_route import ExchangeRoute
+from ...billing.instruments import CreditCard
+from ...utils import is_card_expiring
+
+
+class Routes(object):
```
Comment (dowski) [inline gratipay/models/participant/routes.py]: Would it make more sense to call this `PaymentRoutes` since you also have the concept of routes wiring HTTP requests to code that handles them?

File: gratipay/models/participant/email.py (hunk 1 above)
Comment (dowski) [inline gratipay/models/participant/email.py]: Maybe you can use `not self.has_payout_route`? I think that alone would help.

File: gratipay/models/participant/routes.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    def credit_card_expiring(self):
+        """Return a boolean, whether the participant's credit card is set to expire soon.
+        """
+        route = ExchangeRoute.from_network(self, 'braintree-cc')
+        if not route:
+            return
+        card = CreditCard.from_route(route)
+        year, month = card.expiration_year, card.expiration_month
+        if not (year and month):
+            return False
```
Comment (dowski) [inline gratipay/models/participant/routes.py]: Huh, are there cards without expiration dates?

Comment (whit537) [inline gratipay/models/participant/routes.py]: Dunno, that's pretty old code. 😕

File: gratipay/models/participant/email.py (hunk 2 above)
Comment (whit537) [inline gratipay/models/participant/email.py]: Done in 91a567cdcd69a614b9d26784aba665f5aac63c57.

File: gratipay/models/participant/email.py (hunk 1 above)
Comment (whit537) [inline gratipay/models/participant/email.py]: Done in 91a567cdcd69a614b9d26784aba665f5aac63c57. `has_payout_route` includes failing routes, and we only want non-failing ones here.

File: gratipay/models/participant/routes.py (hunk 6 above)
Comment (whit537) [inline gratipay/models/participant/routes.py]: Fair enough, done in 91a567cdcd69a614b9d26784aba665f5aac63c57. That's a nomenclature wart for sure.

File: gratipay/models/participant/routes.py (hunk 3 above)
Comment (whit537) [inline gratipay/models/participant/routes.py]: Done in 91a567cdcd69a614b9d26784aba665f5aac63c57..

File: gratipay/models/participant/routes.py (hunk 4 above)
Comment (whit537) [inline gratipay/models/participant/routes.py]: Done in 91a567cdcd69a614b9d26784aba665f5aac63c57.

File: gratipay/models/participant/routes.py (hunk 5 above)
Comment (whit537) [inline gratipay/models/participant/routes.py]: Done in 91a567cdcd69a614b9d26784aba665f5aac63c57.

**Comment by whit537** [comment]
Updates in 91a567cdcd69a614b9d26784aba665f5aac63c57, back over to you @dowski. :)
